Improve DNS handling on macOS

pronebird opened the pull request:
## Description

We run a local DNS resolver on loopback interface to workaround Apple Captive Portal check when blocking DNS with firewall. When we don't spoof captive portal, we simply forward traffic to our DNS servers.

Unfortunately, many utilities like `nslookup` or `dig` don't respect custom DNS ports and always use port 53. The original idea to tackle this was by using port forwarding. However PF redirect rules don't seem to work. On top of that they complicate firewall configuration. So the solution is to shift back to port 53 for compatibility.

Since listening directly on `127.0.0.1:53` can cause potential conflicts with other software, I think it's sensible to set an IP alias on loopback interface and use it so that we can listen on port `53`.

The alias is picked randomly from within the `127.1-255.0-255.1-254` range. If after 3 failed attempts the DNS resolver is not able to bind to socket, it will fallback to listening on `127.0.0.1:53`.

Besides that this PR improves DNS reset on macOS 15.x and aligns all versions of macOS by always running DNS resolver. This way we can spot issues a bit earlier too.

## Checklist:

- [x] Changelog

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)
<!-- Reviewable:end -->

pronebird committed:
Fix dns on macOS 15.x

pronebird committed:
Update changelog

pronebird committed:
Patch resolver

pronebird committed:
patch resolver

pronebird committed:
pf: port forwarding is broken so run DNS resolver on port 53

pronebird committed:
Always run local dns resolver on macOS

pronebird committed:
Use random 127.x.x.x ip address

pronebird committed:
Use ioctl to add/remove aliases

pronebird committed:
Update changelog

pronebird committed:
Fix tests

pronebird committed:
clippy: fix

pronebird committed:
Add support for ipv6

pronebird committed:
Tackle ipv6

pronebird committed:
clippy: fix

pronebird committed:
docs: update

pronebird committed:
update changelog

jstuczyn reviewed (changes requested):
> Reviewed 19 of 20 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: 19 of 20 files reviewed, 4 unresolved discussions (waiting on @neacsu)

___
*[`nym-vpn-core/Cargo.toml` line 163 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXSw-b646-tLpUr9SMO:-OXSw-b646-tLpUr9SMP:b-adm14k) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/Cargo.toml#L163)):*
> ```TOML
> tonic = "0.13.1"
> tonic-build = "0.13.1"
> typed-builder = "0.21"
> ```

didn't know about this library - just checked it out, looks pretty cool! (you can obviously dismiss this comment, I just wanted to mention that : D)
___
*[`nym-vpn-core/crates/nym-macos/src/net.rs` line 46 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXSzbR-2xGjxSbEC9Kf:-OXSzbR-2xGjxSbEC9Kg:b-4kr1tc) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-macos/src/net.rs#L46)):*
> ```XML
>             copy_interface_name(interface, &mut req.ifra_name);
> 
>             unsafe { siocaifaddr(ctl_socket.as_raw_fd(), &req as _) }.inspect_err(|e| {
> ```

nit: unsafe always makes me feel a bit uneasy, mind adding a comment on why that is safe to use?
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mod.rs` line 592 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXSyFi_9Kj9B4_unfAP:-OXSyFi_9Kj9B4_unfAQ:b-u8y9fj) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mod.rs#L592)):*
> ```XML
>         #[cfg(target_os = "macos")]
>         {
>             if let Err(e) = self.filtering_resolver_handle.await {
> ```

given that internally the server will wait on `shutdown_gracefully`, maybe to not have the same issue as with the mixnet client, we should introduce some shutdown timeout here? like if the future doesn't resolve in say 10s, we'd abort it
___
*[`nym-vpn-core/crates/nym-firewall/src/macos.rs` line 242 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXSx2AO5WNbf1FvKX9j:-OXSx2AO5WNbf1FvKX9k:bvy2t11) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-firewall/src/macos.rs#L242)):*
> ```XML
>         anchor_change.set_scrub_rules(Self::get_scrub_rules()?);
>         anchor_change.set_filter_rules(new_filter_rules);
>         anchor_change.set_redirect_rules(vec![]);
> ```

reading the docs, setting empty vec will reset the rules as opposed to leave it as it is - are we sure that's what we want? Could it perhaps cause some issues under weird edge cases?
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 75 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXT-cUh4eQCjxk0JVL9:-OXT-cUh4eQCjxk0JVLA:b-mfet5z) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L75)):*
> ```XML
> 
> /// Loopback interface name.
> const LOOPBACK: &str = "lo0";
> ```

I don't think we have to worry about it now, but will hardcoding that value cause problems for some degenerate networking setups where somebody has multiple loopback interfaces?
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 525 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXT0m3X7Tm56CF6t4qb:-OXT0m3X7Tm56CF6t4qc:b806gr4) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L525)):*
> ```XML
>             1u8.max(rand::random()),
>             rand::random(),
>             rand::random::<u8>().clamp(1, 254), // keep last octet in range 1-254
> ```

nit: technically the last byte will be biased towards 1 and 254 since those two values will have double the probability of other options. to avoid it you could use `rng.gen_range()`


<!-- Sent from Reviewable.io -->

neacsu reviewed (commented):
> Reviewed 1 of 20 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: all files reviewed, 6 unresolved discussions (waiting on @pronebird)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 59 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXTZ3QN5pCfDwFmBRkJ:-OXTZ3QN5pCfDwFmBRkK:b-ax459a) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L59)):*
> ```XML
> /// This setting does not affect the error or blocked state. In those states, we will want to use
> /// the local DNS resoler to work around Apple's captive portals check. Exactly how this is done is
> /// documented elsewhere.
> ```

Do you know where that is documented? Maybe it would be useful to link that documentation
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 457 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXTb3yG2iTGzNUuaz1m:-OXTb3yG2iTGzNUuaz1n:bkih3cb) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L457)):*
> ```XML
>             }
> 
>             // SO_REUSEADDR allows us to bind to `127.x.y.z` even if another socket is bound to
> ```

Please double check by testing that we can actually reuse the same port number in this case (with `SO_REUSEADDR` and UDP packets)


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: all files reviewed, 6 unresolved discussions (waiting on @jstuczyn)

___
*[`nym-vpn-core/Cargo.toml` line 163 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXSw-b646-tLpUr9SMO:-OXw4pbwAmFrOe7eEArW:b4c9lyr) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/Cargo.toml#L163)):*
<details><summary><i>Previously, jstuczyn (Jędrzej Stuczyński) wrote…</i></summary><blockquote>

didn't know about this library - just checked it out, looks pretty cool! (you can obviously dismiss this comment, I just wanted to mention that : D)
</blockquote></details>

:thumbsup: 
___
*[`nym-vpn-core/crates/nym-firewall/src/macos.rs` line 242 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXSx2AO5WNbf1FvKX9j:-OXw4wl9B40B4S1DaKFN:b-qtb86s) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-firewall/src/macos.rs#L242)):*
<details><summary><i>Previously, jstuczyn (Jędrzej Stuczyński) wrote…</i></summary><blockquote>

reading the docs, setting empty vec will reset the rules as opposed to leave it as it is - are we sure that's what we want? Could it perhaps cause some issues under weird edge cases?
</blockquote></details>

Yes we indeed want that because we used to have redirect rules, but this is no longer the case. So we want them wiped after upgrade.


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: all files reviewed, 6 unresolved discussions (waiting on @jstuczyn and @neacsu)

___
*[`nym-vpn-core/crates/nym-macos/src/net.rs` line 46 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXSzbR-2xGjxSbEC9Kf:-OXw57rX2Y-MaHmW8Uae:b-denv3j) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-macos/src/net.rs#L46)):*
<details><summary><i>Previously, jstuczyn (Jędrzej Stuczyński) wrote…</i></summary><blockquote>

nit: unsafe always makes me feel a bit uneasy, mind adding a comment on why that is safe to use?
</blockquote></details>

This is a system defined ioctl call from `usr/include/sys/sockio.h`. What do you think I should I put in a safety comment?
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 59 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXTZ3QN5pCfDwFmBRkJ:-OXw6Vfn4VwkzQCP8RKS:b8iejel) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L59)):*
<details><summary><i>Previously, neacsu (Bogdan-Ștefan Neacşu) wrote…</i></summary><blockquote>

Do you know where that is documented? Maybe it would be useful to link that documentation
</blockquote></details>

This comment should be amended because local DNS resolver is used at all times now.
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 525 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXT0m3X7Tm56CF6t4qb:-OXw5hNQ2YeMiqC00v22:b-69q72g) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L525)):*
<details><summary><i>Previously, jstuczyn (Jędrzej Stuczyński) wrote…</i></summary><blockquote>

nit: technically the last byte will be biased towards 1 and 254 since those two values will have double the probability of other options. to avoid it you could use `rng.gen_range()`
</blockquote></details>

Technically this is not very important since we pick an address from such a large range.


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: all files reviewed, 6 unresolved discussions (waiting on @jstuczyn and @neacsu)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 59 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXTZ3QN5pCfDwFmBRkJ:-OXw76Vq3wDCSSfsIAsa:b-5d7c15) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L59)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

This comment should be amended because local DNS resolver is used at all times now.
</blockquote></details>

For captive portal in general here is the document: https://github.com/mullvad/mullvadvpn-app/blob/b54341bb66f5b31aa2a32b89661d9f91a9c11980/docs/allow-macos-network-check.md


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: all files reviewed, 6 unresolved discussions (waiting on @jstuczyn and @neacsu)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mod.rs` line 592 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXSyFi_9Kj9B4_unfAP:-OXwATX54Bu3FoqtLSEY:b-afcnio) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mod.rs#L592)):*
<details><summary><i>Previously, jstuczyn (Jędrzej Stuczyński) wrote…</i></summary><blockquote>

given that internally the server will wait on `shutdown_gracefully`, maybe to not have the same issue as with the mixnet client, we should introduce some shutdown timeout here? like if the future doesn't resolve in say 10s, we'd abort it
</blockquote></details>

Based on my tests this works as expected. Also hickory has a test for graceful shutdown too: https://github.com/hickory-dns/hickory-dns/blob/main/crates/server/src/server/mod.rs#L1040. Let's not introduce arbitrary timeouts until this proves to be a problem.
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 75 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXT-cUh4eQCjxk0JVL9:-OXw7CYmD0bWwdxGZVj9:b-b3dfjy) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L75)):*
<details><summary><i>Previously, jstuczyn (Jędrzej Stuczyński) wrote…</i></summary><blockquote>

I don't think we have to worry about it now, but will hardcoding that value cause problems for some degenerate networking setups where somebody has multiple loopback interfaces?
</blockquote></details>

It should not affect how this works.  `lo0` should always exist on macOS. We configure DNS on `lo0` which is what the system uses for captive portal AFAIK.


<!-- Sent from Reviewable.io -->

pronebird committed:
log -> tracing

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: 11 of 20 files reviewed, 6 unresolved discussions (waiting on @jstuczyn and @neacsu)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 457 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXTb3yG2iTGzNUuaz1m:-OXwZW_f7OqrysZrMzL2:b7ue0ha) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L457)):*
<details><summary><i>Previously, neacsu (Bogdan-Ștefan Neacşu) wrote…</i></summary><blockquote>

Please double check by testing that we can actually reuse the same port number in this case (with `SO_REUSEADDR` and UDP packets)
</blockquote></details>

I wrote a test program:

```rs
/*
* Cargo.toml

[package]
name = "udpbind"
version = "0.1.0"
edition = "2024"

[dependencies]
nix = { version = "0.30.1", features = ["net"] }
anyhow = "*"

*/
use anyhow::Context;
use nix::sys::socket::{self, AddressFamily, SockFlag, SockProtocol, SockType, SockaddrStorage};
use std::{
    net::{IpAddr, Ipv4Addr, SocketAddr, UdpSocket},
    os::fd::AsRawFd,
    thread::{self, JoinHandle},
    time::Duration,
};

const LISTEN_IPS: [IpAddr; 3] = [
    IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)),
    IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
    IpAddr::V4(Ipv4Addr::new(127, 1, 0, 254)),
];

const SEND_TO_IPS: [IpAddr; 2] = [
    IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
    IpAddr::V4(Ipv4Addr::new(127, 1, 0, 254)),
];

const LISTEN_PORT: u16 = 10053;

fn main() -> anyhow::Result<()> {
    // Create interface alias before running this executable:
    // sudo ifconfig lo0 alias 127.1.0.254 255.255.255.0
    // ifconfig lo0

    for reuse_addr_a in [true, false] {
        for addr_a in LISTEN_IPS.iter() {
            for addr_b in LISTEN_IPS.iter() {
                if addr_a == addr_b {
                    continue;
                }

                for reuse_addr_b in [true, false] {
                    let sockaddr_a = SocketAddr::new(*addr_a, LISTEN_PORT);
                    let sockaddr_b = SocketAddr::new(*addr_b, LISTEN_PORT);

                    println!("== Begin test case ==");
                    println!(
                        "{sockaddr_a} {}",
                        if reuse_addr_a {
                            "with SO_REUSEADDR"
                        } else {
                            "without SO_REUSEADDR"
                        }
                    );
                    println!(
                        "{sockaddr_b} {}",
                        if reuse_addr_b {
                            "with SO_REUSEADDR"
                        } else {
                            "without SO_REUSEADDR"
                        }
                    );
                    println!("====================");

                    match run_test_case(sockaddr_a, reuse_addr_a, sockaddr_b, reuse_addr_b) {
                        Ok(()) => println!("Test case passed"),
                        Err(e) => println!("Test case failed: {}", e),
                    }

                    println!("== End test case ==");
                    println!();
                }
            }
        }
    }

    Ok(())
}

fn run_test_case(
    sockaddr_a: SocketAddr,
    reuse_addr_a: bool,
    sockaddr_b: SocketAddr,
    reuse_addr_b: bool,
) -> anyhow::Result<()> {
    let socket_a = create_socket(sockaddr_a, reuse_addr_a)?;
    let socket_b = create_socket(sockaddr_b, reuse_addr_b)?;

    let sockaddr_a = if sockaddr_a.ip().is_unspecified() {
        SocketAddr::new(
            *SEND_TO_IPS
                .iter()
                .find(|ip| **ip != sockaddr_b.ip())
                .unwrap(),
            sockaddr_a.port(),
        )
    } else {
        sockaddr_a
    };
    let sockaddr_b = if sockaddr_b.ip().is_unspecified() {
        SocketAddr::new(
            *SEND_TO_IPS
                .iter()
                .find(|ip| **ip != sockaddr_a.ip())
                .unwrap(),
            sockaddr_b.port(),
        )
    } else {
        sockaddr_b
    };

    send_hello(sockaddr_a, true)
        .with_context(|| format!("Failed to send hello to {sockaddr_a}"))?;
    send_hello(sockaddr_b, false)
        .with_context(|| format!("Failed to send hello to {sockaddr_b}"))?;

    let handle_a = listen_socket(socket_a);
    let handle_b = listen_socket(socket_b);

    handle_a.join().expect("fail to join handle_a");
    handle_b.join().expect("fail to join handle_b");

    Ok(())
}

fn create_socket(listen_addr: SocketAddr, reuse_addr: bool) -> anyhow::Result<UdpSocket> {
    let sock = socket::socket(
        if listen_addr.is_ipv4() {
            AddressFamily::Inet
        } else {
            AddressFamily::Inet6
        },
        SockType::Datagram,
        SockFlag::empty(),
        SockProtocol::Udp,
    )
    .with_context(|| "Failed to open IPv4/UDP socket")?;

    if reuse_addr {
        if let Err(error) = socket::setsockopt(&sock, socket::sockopt::ReuseAddr, &true) {
            eprintln!("Failed to set SO_REUSEADDR on resolver socket: {error}");
        }
    }

    let sin = SockaddrStorage::from(listen_addr);
    socket::bind(sock.as_raw_fd(), &sin)
        .with_context(|| format!("Failed to bind DNS server to {}", listen_addr))?;

    Ok(UdpSocket::from(sock))
}

fn send_hello(ep: SocketAddr, is_a: bool) -> anyhow::Result<()> {
    if is_a {
        println!("Sending hello A to {ep}");
    } else {
        println!("Sending hello B to {ep}");
    }
    let any_addr = SocketAddr::V4("0.0.0.0:0".parse().unwrap());
    let socket =
        UdpSocket::bind(any_addr).with_context(|| format!("Failed to bind socket to {ep}"))?;
    let data: &[u8; 8] = if is_a { b"Hello A!" } else { b"Hello B!" };
    socket
        .send_to(data, ep)
        .with_context(|| format!("Failed to send data to {ep}"))?;
    Ok(())
}

fn listen_socket(socket: UdpSocket) -> JoinHandle<()> {
    let local_addr = socket.local_addr().unwrap();
    thread::spawn(move || {
        let mut buf = [0; 8];
        socket
            .set_read_timeout(Some(Duration::from_secs(2)))
            .expect("failed to set read timeout");
        match socket.recv_from(&mut buf) {
            Ok((amt, src)) => {
                let data = String::from_utf8_lossy(&buf[..amt]);
                println!("{local_addr}: received hello: {data} from {src}");
            }
            Err(error) => {
                eprintln!("{local_addr}: failed to receive data: {error}");
            }
        }
    })
}

```


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: 11 of 20 files reviewed, 6 unresolved discussions (waiting on @jstuczyn and @neacsu)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 457 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXTb3yG2iTGzNUuaz1m:-OXwZdJk8-27AbBqnPe2:b-7mln6y) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L457)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

I wrote a test program:

```rs
/*
* Cargo.toml

[package]
name = "udpbind"
version = "0.1.0"
edition = "2024"

[dependencies]
nix = { version = "0.30.1", features = ["net"] }
anyhow = "*"

*/
use anyhow::Context;
use nix::sys::socket::{self, AddressFamily, SockFlag, SockProtocol, SockType, SockaddrStorage};
use std::{
    net::{IpAddr, Ipv4Addr, SocketAddr, UdpSocket},
    os::fd::AsRawFd,
    thread::{self, JoinHandle},
    time::Duration,
};

const LISTEN_IPS: [IpAddr; 3] = [
    IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)),
    IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
    IpAddr::V4(Ipv4Addr::new(127, 1, 0, 254)),
];

const SEND_TO_IPS: [IpAddr; 2] = [
    IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
    IpAddr::V4(Ipv4Addr::new(127, 1, 0, 254)),
];

const LISTEN_PORT: u16 = 10053;

fn main() -> anyhow::Result<()> {
    // Create interface alias before running this executable:
    // sudo ifconfig lo0 alias 127.1.0.254 255.255.255.0
    // ifconfig lo0

    for reuse_addr_a in [true, false] {
        for addr_a in LISTEN_IPS.iter() {
            for addr_b in LISTEN_IPS.iter() {
                if addr_a == addr_b {
                    continue;
                }

                for reuse_addr_b in [true, false] {
                    let sockaddr_a = SocketAddr::new(*addr_a, LISTEN_PORT);
                    let sockaddr_b = SocketAddr::new(*addr_b, LISTEN_PORT);

                    println!("== Begin test case ==");
                    println!(
                        "{sockaddr_a} {}",
                        if reuse_addr_a {
                            "with SO_REUSEADDR"
                        } else {
                            "without SO_REUSEADDR"
                        }
                    );
                    println!(
                        "{sockaddr_b} {}",
                        if reuse_addr_b {
                            "with SO_REUSEADDR"
                        } else {
                            "without SO_REUSEADDR"
                        }
                    );
                    println!("====================");

                    match run_test_case(sockaddr_a, reuse_addr_a, sockaddr_b, reuse_addr_b) {
                        Ok(()) => println!("Test case passed"),
                        Err(e) => println!("Test case failed: {}", e),
                    }

                    println!("== End test case ==");
                    println!();
                }
            }
        }
    }

    Ok(())
}

fn run_test_case(
    sockaddr_a: SocketAddr,
    reuse_addr_a: bool,
    sockaddr_b: SocketAddr,
    reuse_addr_b: bool,
) -> anyhow::Result<()> {
    let socket_a = create_socket(sockaddr_a, reuse_addr_a)?;
    let socket_b = create_socket(sockaddr_b, reuse_addr_b)?;

    let sockaddr_a = if sockaddr_a.ip().is_unspecified() {
        SocketAddr::new(
            *SEND_TO_IPS
                .iter()
                .find(|ip| **ip != sockaddr_b.ip())
                .unwrap(),
            sockaddr_a.port(),
        )
    } else {
        sockaddr_a
    };
    let sockaddr_b = if sockaddr_b.ip().is_unspecified() {
        SocketAddr::new(
            *SEND_TO_IPS
                .iter()
                .find(|ip| **ip != sockaddr_a.ip())
                .unwrap(),
            sockaddr_b.port(),
        )
    } else {
        sockaddr_b
    };

    send_hello(sockaddr_a, true)
        .with_context(|| format!("Failed to send hello to {sockaddr_a}"))?;
    send_hello(sockaddr_b, false)
        .with_context(|| format!("Failed to send hello to {sockaddr_b}"))?;

    let handle_a = listen_socket(socket_a);
    let handle_b = listen_socket(socket_b);

    handle_a.join().expect("fail to join handle_a");
    handle_b.join().expect("fail to join handle_b");

    Ok(())
}

fn create_socket(listen_addr: SocketAddr, reuse_addr: bool) -> anyhow::Result<UdpSocket> {
    let sock = socket::socket(
        if listen_addr.is_ipv4() {
            AddressFamily::Inet
        } else {
            AddressFamily::Inet6
        },
        SockType::Datagram,
        SockFlag::empty(),
        SockProtocol::Udp,
    )
    .with_context(|| "Failed to open IPv4/UDP socket")?;

    if reuse_addr {
        if let Err(error) = socket::setsockopt(&sock, socket::sockopt::ReuseAddr, &true) {
            eprintln!("Failed to set SO_REUSEADDR on resolver socket: {error}");
        }
    }

    let sin = SockaddrStorage::from(listen_addr);
    socket::bind(sock.as_raw_fd(), &sin)
        .with_context(|| format!("Failed to bind DNS server to {}", listen_addr))?;

    Ok(UdpSocket::from(sock))
}

fn send_hello(ep: SocketAddr, is_a: bool) -> anyhow::Result<()> {
    if is_a {
        println!("Sending hello A to {ep}");
    } else {
        println!("Sending hello B to {ep}");
    }
    let any_addr = SocketAddr::V4("0.0.0.0:0".parse().unwrap());
    let socket =
        UdpSocket::bind(any_addr).with_context(|| format!("Failed to bind socket to {ep}"))?;
    let data: &[u8; 8] = if is_a { b"Hello A!" } else { b"Hello B!" };
    socket
        .send_to(data, ep)
        .with_context(|| format!("Failed to send data to {ep}"))?;
    Ok(())
}

fn listen_socket(socket: UdpSocket) -> JoinHandle<()> {
    let local_addr = socket.local_addr().unwrap();
    thread::spawn(move || {
        let mut buf = [0; 8];
        socket
            .set_read_timeout(Some(Duration::from_secs(2)))
            .expect("failed to set read timeout");
        match socket.recv_from(&mut buf) {
            Ok((amt, src)) => {
                let data = String::from_utf8_lossy(&buf[..amt]);
                println!("{local_addr}: received hello: {data} from {src}");
            }
            Err(error) => {
                eprintln!("{local_addr}: failed to receive data: {error}");
            }
        }
    })
}

```
</blockquote></details>

This is the program output:

```
== Begin test case ==
0.0.0.0:10053 with SO_REUSEADDR
127.0.0.1:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
0.0.0.0:10053: received hello: Hello A! from 127.1.0.254:53372
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:53273
Test case passed
== End test case ==

== Begin test case ==
0.0.0.0:10053 with SO_REUSEADDR
127.0.0.1:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 127.0.0.1:10053
== End test case ==

== Begin test case ==
0.0.0.0:10053 with SO_REUSEADDR
127.1.0.254:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
0.0.0.0:10053: received hello: Hello A! from 127.0.0.1:53753
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:62823
Test case passed
== End test case ==

== Begin test case ==
0.0.0.0:10053 with SO_REUSEADDR
127.1.0.254:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 127.1.0.254:10053
== End test case ==

== Begin test case ==
127.0.0.1:10053 with SO_REUSEADDR
0.0.0.0:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:54962
0.0.0.0:10053: received hello: Hello B! from 127.1.0.254:58730
Test case passed
== End test case ==

== Begin test case ==
127.0.0.1:10053 with SO_REUSEADDR
0.0.0.0:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 0.0.0.0:10053
== End test case ==

== Begin test case ==
127.0.0.1:10053 with SO_REUSEADDR
127.1.0.254:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:54105
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:58056
Test case passed
== End test case ==

== Begin test case ==
127.0.0.1:10053 with SO_REUSEADDR
127.1.0.254:10053 without SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:59202
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:60956
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 with SO_REUSEADDR
0.0.0.0:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:53922
0.0.0.0:10053: received hello: Hello B! from 127.0.0.1:57991
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 with SO_REUSEADDR
0.0.0.0:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 0.0.0.0:10053
== End test case ==

== Begin test case ==
127.1.0.254:10053 with SO_REUSEADDR
127.0.0.1:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:63024
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:62314
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 with SO_REUSEADDR
127.0.0.1:10053 without SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:50911
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:58829
Test case passed
== End test case ==

== Begin test case ==
0.0.0.0:10053 without SO_REUSEADDR
127.0.0.1:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
0.0.0.0:10053: received hello: Hello A! from 127.1.0.254:57351
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:64787
Test case passed
== End test case ==

== Begin test case ==
0.0.0.0:10053 without SO_REUSEADDR
127.0.0.1:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 127.0.0.1:10053
== End test case ==

== Begin test case ==
0.0.0.0:10053 without SO_REUSEADDR
127.1.0.254:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
0.0.0.0:10053: received hello: Hello A! from 127.0.0.1:49963
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:54342
Test case passed
== End test case ==

== Begin test case ==
0.0.0.0:10053 without SO_REUSEADDR
127.1.0.254:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 127.1.0.254:10053
== End test case ==

== Begin test case ==
127.0.0.1:10053 without SO_REUSEADDR
0.0.0.0:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:63081
0.0.0.0:10053: received hello: Hello B! from 127.1.0.254:52033
Test case passed
== End test case ==

== Begin test case ==
127.0.0.1:10053 without SO_REUSEADDR
0.0.0.0:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 0.0.0.0:10053
== End test case ==

== Begin test case ==
127.0.0.1:10053 without SO_REUSEADDR
127.1.0.254:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:55439
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:62738
Test case passed
== End test case ==

== Begin test case ==
127.0.0.1:10053 without SO_REUSEADDR
127.1.0.254:10053 without SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:64453
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:61451
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 without SO_REUSEADDR
0.0.0.0:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:62011
0.0.0.0:10053: received hello: Hello B! from 127.0.0.1:65500
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 without SO_REUSEADDR
0.0.0.0:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 0.0.0.0:10053
== End test case ==

== Begin test case ==
127.1.0.254:10053 without SO_REUSEADDR
127.0.0.1:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:49191
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:62290
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 without SO_REUSEADDR
127.0.0.1:10053 without SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:53252
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:51462
Test case passed
== End test case ==
```


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: 11 of 20 files reviewed, 6 unresolved discussions (waiting on @jstuczyn and @neacsu)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 457 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXTb3yG2iTGzNUuaz1m:-OXwZrQLDhU8Xdc1X9u3:b-vp56nu) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L457)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

This is the program output:

```
== Begin test case ==
0.0.0.0:10053 with SO_REUSEADDR
127.0.0.1:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
0.0.0.0:10053: received hello: Hello A! from 127.1.0.254:53372
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:53273
Test case passed
== End test case ==

== Begin test case ==
0.0.0.0:10053 with SO_REUSEADDR
127.0.0.1:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 127.0.0.1:10053
== End test case ==

== Begin test case ==
0.0.0.0:10053 with SO_REUSEADDR
127.1.0.254:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
0.0.0.0:10053: received hello: Hello A! from 127.0.0.1:53753
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:62823
Test case passed
== End test case ==

== Begin test case ==
0.0.0.0:10053 with SO_REUSEADDR
127.1.0.254:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 127.1.0.254:10053
== End test case ==

== Begin test case ==
127.0.0.1:10053 with SO_REUSEADDR
0.0.0.0:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:54962
0.0.0.0:10053: received hello: Hello B! from 127.1.0.254:58730
Test case passed
== End test case ==

== Begin test case ==
127.0.0.1:10053 with SO_REUSEADDR
0.0.0.0:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 0.0.0.0:10053
== End test case ==

== Begin test case ==
127.0.0.1:10053 with SO_REUSEADDR
127.1.0.254:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:54105
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:58056
Test case passed
== End test case ==

== Begin test case ==
127.0.0.1:10053 with SO_REUSEADDR
127.1.0.254:10053 without SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:59202
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:60956
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 with SO_REUSEADDR
0.0.0.0:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:53922
0.0.0.0:10053: received hello: Hello B! from 127.0.0.1:57991
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 with SO_REUSEADDR
0.0.0.0:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 0.0.0.0:10053
== End test case ==

== Begin test case ==
127.1.0.254:10053 with SO_REUSEADDR
127.0.0.1:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:63024
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:62314
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 with SO_REUSEADDR
127.0.0.1:10053 without SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:50911
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:58829
Test case passed
== End test case ==

== Begin test case ==
0.0.0.0:10053 without SO_REUSEADDR
127.0.0.1:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
0.0.0.0:10053: received hello: Hello A! from 127.1.0.254:57351
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:64787
Test case passed
== End test case ==

== Begin test case ==
0.0.0.0:10053 without SO_REUSEADDR
127.0.0.1:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 127.0.0.1:10053
== End test case ==

== Begin test case ==
0.0.0.0:10053 without SO_REUSEADDR
127.1.0.254:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
0.0.0.0:10053: received hello: Hello A! from 127.0.0.1:49963
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:54342
Test case passed
== End test case ==

== Begin test case ==
0.0.0.0:10053 without SO_REUSEADDR
127.1.0.254:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 127.1.0.254:10053
== End test case ==

== Begin test case ==
127.0.0.1:10053 without SO_REUSEADDR
0.0.0.0:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:63081
0.0.0.0:10053: received hello: Hello B! from 127.1.0.254:52033
Test case passed
== End test case ==

== Begin test case ==
127.0.0.1:10053 without SO_REUSEADDR
0.0.0.0:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 0.0.0.0:10053
== End test case ==

== Begin test case ==
127.0.0.1:10053 without SO_REUSEADDR
127.1.0.254:10053 with SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:55439
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:62738
Test case passed
== End test case ==

== Begin test case ==
127.0.0.1:10053 without SO_REUSEADDR
127.1.0.254:10053 without SO_REUSEADDR
====================
Sending hello A to 127.0.0.1:10053
Sending hello B to 127.1.0.254:10053
127.0.0.1:10053: received hello: Hello A! from 127.0.0.1:64453
127.1.0.254:10053: received hello: Hello B! from 127.1.0.254:61451
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 without SO_REUSEADDR
0.0.0.0:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:62011
0.0.0.0:10053: received hello: Hello B! from 127.0.0.1:65500
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 without SO_REUSEADDR
0.0.0.0:10053 without SO_REUSEADDR
====================
Test case failed: Failed to bind DNS server to 0.0.0.0:10053
== End test case ==

== Begin test case ==
127.1.0.254:10053 without SO_REUSEADDR
127.0.0.1:10053 with SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:49191
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:62290
Test case passed
== End test case ==

== Begin test case ==
127.1.0.254:10053 without SO_REUSEADDR
127.0.0.1:10053 without SO_REUSEADDR
====================
Sending hello A to 127.1.0.254:10053
Sending hello B to 127.0.0.1:10053
127.0.0.1:10053: received hello: Hello B! from 127.0.0.1:53252
127.1.0.254:10053: received hello: Hello A! from 127.1.0.254:51462
Test case passed
== End test case ==
```
</blockquote></details>

Note that sending to `0.0.0.0:10053` is equivalent to sending it to `127.0.0.1:10053` so I am swapping `0.0.0.0` for the other IP that is not used in the test case, alternating between `127.0.0.1` and `127.1.0.254` (`lo0` alias)


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: 11 of 20 files reviewed, 6 unresolved discussions (waiting on @jstuczyn and @neacsu)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 457 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXTb3yG2iTGzNUuaz1m:-OXwaDcL6ayQ9XRRo6f-:bni7mn4) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L457)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Note that sending to `0.0.0.0:10053` is equivalent to sending it to `127.0.0.1:10053` so I am swapping `0.0.0.0` for the other IP that is not used in the test case, alternating between `127.0.0.1` and `127.1.0.254` (`lo0` alias)
</blockquote></details>

Gist: https://gist.github.com/pronebird/32270037d2e6cf801761e51e4e26b7c5


<!-- Sent from Reviewable.io -->

pronebird committed:
Update doc comment

pronebird committed:
Do not ignore loopback addrs when setting DNS

Related change: https://github.com/mullvad/mullvadvpn-app/commits/main/talpid-core/src/dns/macos.rs

pronebird committed:
Improve range gen

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: 10 of 21 files reviewed, 6 unresolved discussions (waiting on @jstuczyn and @neacsu)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs` line 525 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232#-OXT0m3X7Tm56CF6t4qb:-OXwsLuX1vroUHfaeiqy:b-dem6e) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/5ac64f2038489291fb6af323cac12e52f1a26e8a/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/resolver.rs#L525)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Technically this is not very important since we pick an address from such a large range.
</blockquote></details>

Alright let's improve that with `thread_rng().gen_range()`, seems like an easy fix


<!-- Sent from Reviewable.io -->

neacsu reviewed (commented):
> Reviewed 9 of 11 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3232)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->
